### PR TITLE
[Joox] Add new extractor

### PR DESCRIPTION
**Pull request by HugoTai100**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
A new extractor for www.joox.com

### CODE REVIEW

File: youtube_dl/extractor/joox.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            'ext': 'mp3',
+            'title': '忍',
+            'release_date': '20180214',
+            'creator': '林欣彤',
+            'thumbnail': r're:^https?://.*\.jpg$',
+        }
+    }]
+
+    def _real_extract(self, url):
+        _VALID_URL = r'https?://www\.joox\.com/(?P<country>[a-z]*)/(?P<language>[a-z]*_?[a-z]*)/single/(?P<id>[a-zA-z0-9+]*==)'
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: Code duplication.

File: youtube_dl/extractor/joox.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            'songid': song_id,
+            'lang': _lang,
+            'country': _country,
+            'form_type': -1,
+            'channel_id': -1,
+            '_': _code
+        }
+
+        detail_info_page = self._download_webpage(
+            "http://api.joox.com/web-fcgi-bin/web_get_songinfo?" + compat_urllib_parse_urlencode(query), song_id)
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: `query=` parameter.

File: youtube_dl/extractor/joox.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        singer = singer[0].get('name')
+        singer = base64.b64decode(singer).decode('UTF-8')
+        publish_time = song_json.get('public_time')
+        publish_time = publish_time.replace('-', '')
+        formats = []
+        _FORMATS = {
+            '128m4a': {'url': song192mp3, 'abr': 128},
+            'm4a': {'url': songm4a},
+            'mp3': {'url': songmp3, 'abr': 128, 'filesize': size128},
+            '320mp3': {'url': song320mp3, 'preference': -1, 'abr': 320, 'filesize': size320},
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: URLs must not be `None`.

File: youtube_dl/extractor/joox.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        for format_id, details in _FORMATS.items():
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: This intermediate dict is completely pointless. Build formats directly.

File: youtube_dl/extractor/joox.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            formats.append({
+                'url': details['url'],
+                'format': format_id,
+                'format_id': format_id,
+                'preference': details.get('preference'),
+                'abr': details.get('abr'),
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: Must be int.

File: youtube_dl/extractor/joox.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                'filesize': details.get('filesize'),
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: Must be int.

File: youtube_dl/extractor/joox.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                'resolution': 'audio only'
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: Remove.

File: youtube_dl/extractor/joox.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        _FORMATS = {
+            '128m4a': {'url': song192mp3, 'abr': 128},
+            'm4a': {'url': songm4a},
+            'mp3': {'url': songmp3, 'abr': 128, 'filesize': size128},
+            '320mp3': {'url': song320mp3, 'preference': -1, 'abr': 320, 'filesize': size320},
+        }
+        for format_id, details in _FORMATS.items():
+            formats.append({
+                'url': details['url'],
+                'format': format_id,
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: Remove.

File: youtube_dl/extractor/joox.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                'format_id': format_id,
+                'preference': details.get('preference'),
+                'abr': details.get('abr'),
+                'filesize': details.get('filesize'),
+                'resolution': 'audio only'
+            })
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: `'vcodec': 'none'`.

File: youtube_dl/extractor/joox.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        return {
+            'id': song_id,
+            'title': song_title,
+            'formats': formats,
+            'thumbnail': album_thumbnail,
+            'release_date': publish_time,
+            'duration': _duration,
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: Must be int.

File: youtube_dl/extractor/joox.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            'creator': singer,
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: `artist`.

File: youtube_dl/extractor/joox.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        detail_info_page = self._download_webpage(
+            "http://api.joox.com/web-fcgi-bin/web_get_songinfo?" + compat_urllib_parse_urlencode(query), song_id)
+        detail_info_page = detail_info_page[18:-1]
+        song_json = self._parse_json(detail_info_page, song_id)
+        song320mp3 = song_json.get('r320Url')
+        song192mp3 = song_json.get('r192Url')
+        songmp3 = song_json.get('mp3Url')
+        songm4a = song_json.get('m4aUrl')
+        song_title = song_json.get('msong')
+        _duration = song_json.get('minterval')
```
Comment (dstftw) [inline youtube_dl/extractor/joox.py]: Do not use underscore as prefix.

**Comment by HugoTai100** [comment]
I have added the changed accordingly.

**Comment by dstftw** [comment]
Re-read my comments, re-read coding conventions and fix all remaining issues.

**Pull request closed without merging**